### ISSUE TITLE
Profiler with schedule only prints one metrics

### ISSUE THREAD

**Issue by FrancescoSaverioZuppichini** [pytorch/pytorch#121141]
### 🐛 Describe the bug

Running the profiler with `schedule` results in the following table to be printed

```python
import torch
import torchvision.models as models
from torch.profiler import profile, record_function, ProfilerActivity
import os
from src.utils import get_logger, get_torch_profiler, schedule

os.environ["TORCH_COMPILE_DEBUG"] = "1"
batch_size = [1, 4, 8, 16, 32, 64]
model = models.resnet18()
model = model.cuda()

logger = get_logger()


def run(batch_size: int):
    inputs = (torch.rand((batch_size, 3, 224, 224)) * 255).to(dtype=torch.uint8)
    with profile(schedule=schedule(wait=0, warmup=8, repeat=4, active=4)) as prof:
        with record_function("to gpu"):
            inputs = inputs.cuda()
            inputs = inputs.to(torch.float32)
        with record_function("to float"):
            inputs = inputs.to(torch.float32)
        with record_function("forward"):
            model(inputs)

    prof.export_chrome_trace("trace.json")
    keys = prof.key_averages()
    print(prof.key_averages().table(row_limit=10))
    my_keys = list(filter(lambda e: e.key in ["to gpu", "to float", "forward"], keys))
    print(prof.key_averages().table(row_limit=10, top_level_events_only=True))


run(1)

```

```
(.venv) (base) ➜  py4ai git:(main) ✗ /home/zuppif/Documents/py4ai/.venv/bin/python /home/zuppif/Documents/py4ai/playground.py
STAGE:2024-03-04 17:49:56 79891:79891 ActivityProfilerController.cpp:314] Completed Stage: Warm Up
STAGE:2024-03-04 17:49:56 79891:79891 ActivityProfilerController.cpp:320] Completed Stage: Collection
STAGE:2024-03-04 17:49:56 79891:79891 ActivityProfilerController.cpp:324] Completed Stage: Post Processing
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                     Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
    cudaDeviceSynchronize       100.00%      16.000us       100.00%      16.000us      16.000us             1  
-------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 16.000us
```

with is useless because I don't see my record function, If I remove `schedule`, I get what I want

```
STAGE:2024-03-04 17:50:42 79993:79993 ActivityProfilerController.cpp:314] Completed Stage: Warm Up
STAGE:2024-03-04 17:50:42 79993:79993 ActivityProfilerController.cpp:320] Completed Stage: Collection
STAGE:2024-03-04 17:50:42 79993:79993 ActivityProfilerController.cpp:324] Completed Stage: Post Processing
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                 to gpu         0.10%     242.000us         6.94%      16.259ms      16.259ms       0.000us         0.00%      24.000us      24.000us             1  
                                               aten::to         0.76%       1.782ms         6.84%      16.017ms       5.339ms       0.000us         0.00%      24.000us       8.000us             3  
                                         aten::_to_copy         0.01%      31.000us         6.08%      14.235ms       7.117ms       0.000us         0.00%      24.000us      12.000us             2  
                                    aten::empty_strided         0.02%      44.000us         0.09%     200.000us     100.000us       0.000us         0.00%       7.000us       3.500us             2  
                                            aten::copy_         0.03%      62.000us         5.98%      14.004ms       7.002ms      17.000us         2.41%      17.000us       8.500us             2  
                                        cudaMemcpyAsync         0.02%      51.000us         0.02%      51.000us      51.000us       0.000us         0.00%       0.000us       0.000us             1  
                       Memcpy HtoD (Pageable -> Device)         0.00%       0.000us         0.00%       0.000us       0.000us      13.000us         1.84%      13.000us      13.000us             1  
                                  cudaStreamSynchronize         0.01%      15.000us         0.01%      15.000us      15.000us       0.000us         0.00%       0.000us       0.000us             1  
                                  cudaStreamIsCapturing         0.01%      16.000us         0.01%      16.000us       0.333us       7.000us         0.99%       7.000us       0.146us            48  
                                             cudaMalloc         0.62%       1.448ms         0.62%       1.448ms      96.533us       0.000us         0.00%       0.000us       0.000us            15  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 234.306ms
Self CUDA time total: 706.000us
```

Thanks a lot

### Versions

```
PyTorch version: 2.2.1+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.12.2 | packaged by Anaconda, Inc. | (main, Feb 27 2024, 17:35:02) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.5.0-21-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3090
Nvidia driver version: 535.161.07
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      43 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             16
On-line CPU(s) list:                0-15
Vendor ID:                          AuthenticAMD
Model name:                         AMD Ryzen 7 2700X Eight-Core Processor
CPU family:                         23
Model:                              8
Thread(s) per core:                 2
Core(s) per socket:                 8
Socket(s):                          1
Stepping:                           2
Frequency boost:                    enabled
CPU max MHz:                        3700,0000
CPU min MHz:                        2200,0000
BogoMIPS:                           7399.01
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb hw_pstate ssbd ibpb vmmcall fsgsbase bmi1 avx2 smep bmi2 rdseed adx smap clflushopt sha_ni xsaveopt xsavec xgetbv1 clzero irperf xsaveerptr arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif overflow_recov succor smca sev sev_es
Virtualization:                     AMD-V
L1d cache:                          256 KiB (8 instances)
L1i cache:                          512 KiB (8 instances)
L2 cache:                           4 MiB (8 instances)
L3 cache:                           16 MiB (2 instances)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-15
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Mitigation; untrained return thunk; SMT vulnerable
Vulnerability Spec rstack overflow: Mitigation; safe RET
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, IBPB conditional, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] torch==2.2.1
[pip3] torchaudio==2.2.1
[pip3] torchvision==0.17.1
[conda] _anaconda_depends         2023.09             py311_mkl_1  
[conda] blas                      1.0                         mkl  
[conda] mkl                       2023.1.0         h213fc3f_46344  
[conda] mkl-service               2.4.0           py311h5eee18b_1  
[conda] mkl_fft                   1.3.8           py311h5eee18b_0  
[conda] mkl_random                1.2.4           py311hdb19cb5_0  
[conda] numpy                     1.26.3          py311h08b1b3b_0  
[conda] numpy-base                1.26.3          py311hf175353_0  
[conda] numpydoc                  1.5.0           py311h06a4308_0 
```

cc @robieta @chaekit @aaronenyeshi @nbcsm @guotuofeng @guyang3532 @gaoteng-git @tiffzhaofb @dzhulgakov @davidberard98

**Comment by sraikund16** [comment]
Hello, it looks like the issue is based on the schedule you are using and the number of steps you are running. In the example above, you are only running for a single iteration ,but your warmup is for 8 steps. Because of this, you don't get any of the metrics you expect since the state never gets to the `RECORD` state. You can either set warmup to 0, or add a `for loop` to add more steps. Please refer to this page for an example: https://pytorch.org/tutorials/recipes/recipes/profiler_recipe.html

**Comment by FrancescoSaverioZuppichini** [comment]
> Hello, it looks like the issue is based on the schedule you are using and the number of steps you are running. In the example above, you are only running for a single iteration, but your warmup is for 8 steps. Because of this, you don't get any of the metrics you expect since the profiler never gets to the `RECORD` state. You can either set warmup to 0, or add a `for loop` to add more steps. Please refer to this page for an example: https://pytorch.org/tutorials/recipes/recipes/profiler_recipe.html

thanks a lot, I am wondering if I want to `warmup` for `X` steps and then profiling to `Y` steps, should I do `warmup=X, steps=X+Y`?

**Comment by FrancescoSaverioZuppichini** [comment]
Thanks a lot, will try and report back :)

**Comment by FrancescoSaverioZuppichini** [comment]
Not sure I got it how it works

```python
def get_torch_profiler() -> profile:
    return profile(
        activities=[ProfilerActivity.CPU, ProfilerActivity.CUDA],
        schedule=schedule(wait=0, warmup=4, active=6),
    )
...

 with get_torch_profiler() as prof:
      with record_function("run_dataloader"):
          for _ in tqdm(range(num_epoches)):
              for batch in dl:
                 continue
                   
                        
```

where `num_epoches` is `10`, just return one metrics

Thank you so much

**Comment by sraikund16** [comment]
Try using the example at the end of the page of the link I posted in my previous comment. You need to use the `step` function of the profiler so that the state eventually enters RECORD

**Comment by FrancescoSaverioZuppichini** [comment]
> Try using the example at the end of the page in the link I posted in my previous comment. You need to use the `step` function of the profiler so that the state eventually enters RECORD

Doing

```python
my_schedule = schedule(
    skip_first=0,
    wait=1,
    warmup=4,
    active=5)

def trace_handler(p):
    output = p.key_averages().table(sort_by="self_cuda_time_total", row_limit=10)
    print("trace_handler")
    print(output)
    p.export_chrome_trace("trace_" + str(p.step_num) + ".json")

with profile(activities=[ProfilerActivity.CPU, ProfilerActivity.CUDA], schedule=my_schedule) as prof:
    for _ in range(10):
        with record_function("model_inference"):
            model(inputs)
        prof.step()
```

Doesn't even print inside the handler ...

**Comment by aaronenyeshi** [comment]
> Doesn't even print inside the handler ...

You need to add the handler so that it gets called when traces are ready. try running this code:
```
my_schedule = schedule(
    wait=1,
    warmup=4,
    active=5,
    repeat=1)

 def trace_handler(p):
    output = p.key_averages().table(sort_by="self_cuda_time_total", row_limit=10)
    print("trace_handler")
    print(output)
    p.export_chrome_trace("trace_" + str(p.step_num) + ".json")

with profile(
    activities=[ProfilerActivity.CPU, ProfilerActivity.CUDA],
    schedule=my_schedule),
    on_trace_ready=trace_handler,
as prof:
    for _ in range(10):
        with record_function("model_inference"):
            model(inputs)
        prof.step()
```

**Comment by sraikund16** [comment]
Tested locally, this is a user issue.
